README.md
coreboot README
===============

coreboot is a Free Software project aimed at replacing the proprietary BIOS
(firmware) found in most computers.  coreboot performs a little bit of
hardware initialization and then executes additional boot logic, called a
payload.

With the separation of hardware initialization and later boot logic,
coreboot can scale from specialized applications that run directly
firmware, run operating systems in flash, load custom
bootloaders, or implement firmware standards, like PC BIOS services or
UEFI. This allows for systems to only include the features necessary
in the target application, reducing the amount of code and flash space
required.

coreboot was formerly known as LinuxBIOS.


Payloads
--------

After the basic initialization of the hardware has been performed, any
desired "payload" can be started by coreboot.

See <https://www.coreboot.org/Payloads> for a list of supported payloads.


Supported Hardware
------------------

coreboot supports a wide range of chipsets, devices, and mainboards.

For details please consult:

 * <https://www.coreboot.org/Supported_Motherboards>


Build Requirements
------------------

 * make
 * gcc / g++
   Because Linux distribution compilers tend to use lots of patches. coreboot
   does lots of "unusual" things in its build system, some of which break due
   to those patches, sometimes by gcc aborting, sometimes - and that's worse -
   by generating broken object code.
   Two options: use our toolchain (eg. make crosstools-i386) or enable the
   `ANY_TOOLCHAIN` Kconfig option if you're feeling lucky (no support in this
   case).
 * iasl (for targets with ACPI support)
 * pkg-config
 * libssl-dev (openssl)

Optional:

 * doxygen (for generating/viewing documentation)
 * gdb (for better debugging facilities on some targets)
 * ncurses (for `make menuconfig` and `make nconfig`)
 * flex and bison (for regenerating parsers)


Building coreboot
-----------------

Please consult <https://www.coreboot.org/Build_HOWTO> for details.


Testing coreboot Without Modifying Your Hardware
------------------------------------------------

If you want to test coreboot without any risks before you really decide
to use it on your hardware, you can use the QEMU system emulator to run
coreboot virtually in QEMU.

Please see <https://www.coreboot.org/QEMU> for details.


Website and Mailing List
------------------------

Further details on the project, a FAQ, many HOWTOs, news, development
guidelines and more can be found on the coreboot website:

  <https://www.coreboot.org>

You can contact us directly on the coreboot mailing list:

  <https://www.coreboot.org/Mailinglist>


Copyright and License
---------------------

The copyright on coreboot is owned by quite a large number of individual
developers and companies. Please check the individual source files for details.

coreboot is licensed under the terms of the GNU General Public License (GPL).
Some files are licensed under the "GPL (version 2, or any later version)",
and some files are licensed under the "GPL, version 2". For some parts, which
were derived from other projects, other (GPL-compatible) licenses may apply.
Please check the individual source files for details.

This makes the resulting coreboot images licensed under the GPL, version 2.


Issues
-----------

If you have any trouble or find any bug, please report an issue in 
[this location](https://github.com/pcengines/apu2-documentation/issues). To 
create the issue, from the list of available templates select the one, that fits 
best the nature of the issue (bug, feature, question or task) and fill it.


CHANGELOG.md
Change log for PC Engines fork of coreboot
==========================================

Releases 4.0.x are based on PC Engines 20160304 release.
Releases 4.5.x and 4.6.x are based on mainline support submitted in
[this gerrit ref](https://review.coreboot.org/#/c/14138/).
Releases 4.8.0.x/4.9.0.x and later are based on continuous synchronization with
official [coreboot repository](https://review.coreboot.org/cgit/coreboot.git)

## Quick build instructions

Please use [pce-fw-builder](https://github.com/pcengines/pce-fw-builder)
****
## [Unreleased]

## [v4.16.0.4] - 2022-05-19
### Changed
- Rebased with official coreboot repository commit 9686ac2261
- [updated sortbootorder to v4.6.23](https://github.com/pcengines/sortbootorder/blob/master/CHANGELOG.md#v4623---2022-05-19)
- [updated SeaBIOS to rel-1.16.0.1](https://github.com/pcengines/seabios/blob/apu_support/CHANGELOG.md#rel-11601---2022-05-19)

## [v4.16.0.3] - 2022-04-21
### Changed
- Rebased with official coreboot repository commit 2c4b426557

## [v4.16.0.2] - 2022-03-29
### Changed
- Rebased with official coreboot repository commit 66f99f7f

## [v4.16.0.1] - 2022-02-11
### Changed
- Rebased with official coreboot repository commit b4ba289f
- Disabled loglevel prefixes introduced in coreboot 4.16
- Disabled ANSI escape sequences introduced in coreboot 4.16

### Fixed
- [AMD PSP CCP as entropy source](https://github.com/pcengines/apu2-documentation/issues/112)

## [v4.15.0.3] - 2022-02-11
### Changed
- Rebased with official coreboot repository commit 36425312ee
- Added checking hardware matrix before regression tests

### Fixed
- The hard disk not visible in the Seabios Boot Menu

## [v4.15.0.2] - 2021-12-27
### Changed
- rebased with official coreboot repository commit 3990da0b
- disabled SMM
- enabled parallel AP initialization for apu2-6 for faster boot time

## [v4.15.0.1] - 2021-11-26
### Changed
- rebased with official coreboot repository commit 6973a3e7

## [v4.14.0.6] - 2021-10-29
### Changed
- rebased with official coreboot repository commit d06c0917

## [v4.14.0.5] - 2021-10-08
### Changed
- rebased with official coreboot repository commit d4c55353
- Updated CPU declarations in ACPI to comply with newer ACPI standard
- Removed GPIO bindings to fix conflict with OS drivers

## [v4.14.0.4] - 2021-09-06
### Changed
- rebased with official coreboot repository commit d9f5d90
- enabled EHCI controller by default on apu3-apu6 platforms
- [updated sortbootorder to v4.6.22](https://github.com/pcengines/sortbootorder/blob/master/CHANGELOG.md#v4622---2021-09-06)

### Added
- safeguard against setting watchdog timeout too low

## [v4.14.0.3] - 2021-07-29
### Changed
- rebased with official coreboot repository commit c049c80

## [v4.14.0.2] - 2021-06-28
### Changed
- rebased with official coreboot repository commit 1c43d92

## [v4.14.0.1] - 2021-05-27
### Changed
- rebased with official coreboot repository commit 6a936fc
- [updated sortbootorder to v4.6.21](https://github.com/pcengines/sortbootorder/blob/master/CHANGELOG.md#v4621---2021-05-27)
- [updated SeaBIOS to rel-1.14.0.1](https://github.com/pcengines/seabios/blob/apu_support/CHANGELOG.md#rel-11401---2021-05-27)

### Added
- persistent bootorder and runtime configuration in separate FMAP region for
  apu2-apu6 platforms

### Fixed
- inconsistent MTRRs warning in Linux

## [v4.13.0.6] - 2021-04-29
### Changed
- rebased with official coreboot repository commit a4c09c5

## [v4.13.0.5] - 2021-04-07
### Changed
- rebased with official coreboot repository commit e7a68ec

## [v4.13.0.4] - 2021-02-24
### Changed
- rebased with official coreboot repository commit c79fc47

## [v4.13.0.3] - 2021-01-27
### Changed
- rebased with official coreboot repository commit 5e6e5c1

## [v4.13.0.2] - 2020-12-28
### Changed
- rebased with official coreboot repository commit 8edb48b

### Fixed
- [Serial number calculation on APU1](https://github.com/pcengines/coreboot/issues/436)

## [v4.13.0.1] - 2020-11-25
### Changed
- rebased with official coreboot repository commit 9b7dc76

## [v4.12.0.6] - 2020-10-29
### Changed
- rebased with official coreboot repository commit 43439f6

### Fixed
- the option in runtime config to reverse PCI addressing order, now not only
  mPCIe devices, but NICs are reversed as well. The WoL capable NIC should be
  the first booting in iPXE when the reverse option is enabled.

## [v4.12.0.5] - 2020-09-25
### Changed
- rebased with official coreboot repository commit da3375e
- added apu6 build target

## [v4.12.0.4] - 2020-08-28
### Changed
- rebased with official coreboot repository commit 81a2f45
- updates sortbootorder to v4.6.20 for minor build fix

### Fixed
- TPM2 visibility in OS for apu3d and apu4d
- issues with IRQ vectors reported in Xen and Linux dmesg

## [v4.12.0.3] - 2020-07-29
### Fixed
- [memory speed values in the DMI table](https://github.com/pcengines/apu2-documentation/issues/176)
- [Cbmem error on apu1 caused by wrong ACPI table](https://github.com/pcengines/coreboot/issues/411)

### Changed
- rebased with official coreboot repository commit 8d5cedf
- [updated sortbootorder to v4.6.19](https://github.com/pcengines/sortbootorder/blob/master/CHANGELOG.md#v4619---2020-07-29)
- [extended flashing documentation with FreeBSD/pfSense guides](https://github.com/pcengines/apu2-documentation/blob/master/docs/firmware_flashing.md)

### Added
- [option in runtime config to reverse PCI addressing order](https://github.com/pcengines/coreboot/issues/392)

## [v4.12.0.2] - 2020-06-28
### Fixed
- [incorrrect serial number in dmidecode for apu1](https://github.com/pcengines/coreboot/issues/402)

### Changed
- rebased with official coreboot repository commit f183626
- enabled TPM2 for apu3d and apu4d supporting a TPM module connection

### Added
- DRTM ACPI table (holds information about TPM log area for TrenchBoot DRTM)

## [v4.12.0.1] - 2020-05-29
### Changed
- rebased with official coreboot repository commit 8952d1c

src/drivers/net/r8168.c
/* SPDX-License-Identifier: GPL-2.0-only */

/*
 * This driver resets the 10ec:8168 NIC then tries to read
 * "macaddress" string XX:XX:XX:XX:XX:XX from CBFS.
 * If no MAC is found, it programs a default MAC address in the device
 * so that if the EEPROM/efuse is unconfigured it still has a default MAC.
 */

#include <cbfs.h>
#include <acpi/acpi_device.h>
#include <acpi/acpigen.h>
#include <string.h>
#include <arch/io.h>
#include <console/console.h>
#include <device/device.h>
#include <device/pci.h>
#include <device/pci_ops.h>
#include <device/pci_def.h>
#include <device/pci_ids.h>
#include <delay.h>
#include <fmap.h>
#include <types.h>

#include "chip.h"

#define NIC_TIMEOUT		1000

#define CMD_REG			0x37
#define  CMD_REG_RESET		0x10
#define CMD_LED0_LED1		0x18
#define CMD_LED_FEATURE		0x94
#define CMD_LEDSEL0		0x18
#define CMD_LEDSEL2		0x84

#define CFG_9346		0x50
#define  CFG_9346_LOCK		0x00
#define  CFG_9346_UNLOCK	0xc0
#define CMD_REG_ASPM		0xb0
#define ASPM_L1_2_MASK		0xe059000f

#define DEVICE_INDEX_BYTE	12
#define MAX_DEVICE_SUPPORT	10
/**
 * search: Find first instance of string in a given region
 * @param p       string to find
 * @param a       start address of region to search
 * @param lengthp length of string to search for
 * @param lengtha length of region to search in
 * @return offset offset from start address a where string was found.
 *                If string not found, return lengtha.
 */
static size_t search(const char *p, const u8 *a, size_t lengthp,
		     size_t lengtha)
{
	size_t i, j;

	if (lengtha < lengthp)
		return lengtha;
	/* Searching */
	for (j = 0; j <= lengtha - lengthp; j++) {
		for (i = 0; i < lengthp && p[i] == a[i + j]; i++)
			;
		if (i >= lengthp && a[j - 1] == lengthp)
			return j;
	}
	return lengtha;
}

static u8 get_hex_digit(const u8 c)
{
	u8 ret = 0;

	ret = c - '0';
	if (ret > 0x09) {
		ret = c - 'A' + 0x0a;
		if (ret > 0x0f)
			ret = c - 'a' + 0x0a;
	}
	if (ret > 0x0f) {
		printk(BIOS_ERR, "Invalid hex digit found: "
				 "%c - 0x%02x\n", (char)c, c);
		ret = 0;
	}
	return ret;
}

#define MACLEN 17

/* Returns MAC address based on the key that is passed in. */
static enum cb_err fetch_mac_vpd_key(u8 *macstrbuf, const char *vpd_key)
{
	struct region_device rdev;
	void *search_address;
	size_t search_length;
	size_t offset;

	if (fmap_locate_area_as_rdev("RO_VPD", &rdev)) {
		printk(BIOS_ERR, "Couldn't find RO_VPD region.");
		return CB_ERR;
	}
	search_address = rdev_mmap_full(&rdev);
	if (search_address == NULL) {
		printk(BIOS_ERR, "LAN: VPD not found.\n");
		return CB_ERR;
	}

	search_length = region_device_sz(&rdev);
	offset = search(vpd_key, search_address, strlen(vpd_key),
			search_length);

	if (offset == search_length) {
		printk(BIOS_ERR, "Could not locate '%s' in VPD\n", vpd_key);
		rdev_munmap(&rdev, search_address);
		return CB_ERR;
	}
	printk(BIOS_DEBUG, "Located '%s' in VPD\n", vpd_key);

	offset += strlen(vpd_key) + 1;	/* move to next character */

	if (offset + MACLEN > search_length) {
		rdev_munmap(&rdev, search_address);
		printk(BIOS_ERR, "Search result too small!\n");
		return CB_ERR;
	}
	memcpy(macstrbuf, search_address + offset, MACLEN);
	rdev_munmap(&rdev, search_address);

	return CB_SUCCESS;
}

/* Prepares vpd_key by concatenating ethernet_mac with device_index */
static enum cb_err fetch_mac_vpd_dev_idx(u8 *macstrbuf, u8 device_index)
{
	char key[] = "ethernet_mac "; /* Leave a space at tail to stuff an index */

	/*
	 * Map each NIC on the DUT to "ethernet_macN", where N is [0-9].
	 * Translate index number from integer to ascii by adding '0' char.
	 */
	key[DEVICE_INDEX_BYTE] = device_index + '0';

	return fetch_mac_vpd_key(macstrbuf, key);
}

static void fetch_mac_string_vpd(struct drivers_net_config *config, u8 *macstrbuf)
{
	if (!config)
		return;

	/* Current implementation is up to 10 NIC cards */
	if (config->device_index > MAX_DEVICE_SUPPORT) {
		printk(BIOS_ERR, "r8168: the maximum device_index should be less then %d\n."
					" Using default 00:e0:4c:00:c0:b0\n", MAX_DEVICE_SUPPORT);
		return;
	}

	if (fetch_mac_vpd_dev_idx(macstrbuf, config->device_index) == CB_SUCCESS)
		return;

	if (!CONFIG(RT8168_SUPPORT_LEGACY_VPD_MAC)) {
		printk(BIOS_ERR, "r8168: mac address not found in VPD,"
						 " using default 00:e0:4c:00:c0:b0\n");
		return;
	}

	if (fetch_mac_vpd_key(macstrbuf, "ethernet_mac") != CB_SUCCESS)
		printk(BIOS_ERR, "r8168: mac address not found in VPD,"
					 " using default 00:e0:4c:00:c0:b0\n");
}

static enum cb_err fetch_mac_string_cbfs(u8 *macstrbuf)
{
	if (!cbfs_load("rt8168-macaddress", macstrbuf, MACLEN)) {
		printk(BIOS_ERR, "r8168: Error reading MAC from CBFS\n");
		return CB_ERR;
	}
	return CB_SUCCESS;
}

static void get_mac_address(u8 *macaddr, const u8 *strbuf)
{
	size_t offset = 0;
	int i;

	if ((strbuf[2] != ':') || (strbuf[5] != ':') ||
	    (strbuf[8] != ':') || (strbuf[11] != ':') ||
	    (strbuf[14] != ':')) {
		printk(BIOS_ERR, "r8168: ignore invalid MAC address in cbfs\n");
		return;
	}

	for (i = 0; i < 6; i++) {
		macaddr[i] = 0;
		macaddr[i] |= get_hex_digit(strbuf[offset]) << 4;
		macaddr[i] |= get_hex_digit(strbuf[offset + 1]);
		offset += 3;
	}
}

static void program_mac_address(struct device *dev, u16 io_base)
{
	u8 macstrbuf[MACLEN] = { 0 };
	int i = 0;
	/* Default MAC Address of 00:E0:4C:00:C0:B0 */
	u8 mac[6] = { 0x00, 0xe0, 0x4c, 0x00, 0xc0, 0xb0 };
	struct drivers_net_config *config = dev->chip_info;

	/* check the VPD for the mac address */
	if (CONFIG(RT8168_GET_MAC_FROM_VPD)) {
		fetch_mac_string_vpd(config, macstrbuf);
	} else {
		if (fetch_mac_string_cbfs(macstrbuf) != CB_SUCCESS)
			printk(BIOS_ERR, "r8168: Error reading MAC from CBFS,"
							" using default 00:e0:4c:00:c0:b0\n");
	}
	get_mac_address(mac, macstrbuf);

	/* Reset NIC */
	printk(BIOS_DEBUG, "r8168: Resetting NIC...");
	outb(CMD_REG_RESET, io_base + CMD_REG);

	/* Poll for reset, with 1sec timeout */
	while (i < NIC_TIMEOUT && (inb(io_base + CMD_REG) & CMD_REG_RESET)) {
		udelay(1000);
		if (++i >= NIC_TIMEOUT)
			printk(BIOS_ERR, "timeout waiting for nic to reset\n");
	}
	if (i < NIC_TIMEOUT)
		printk(BIOS_DEBUG, "done\n");

	printk(BIOS_DEBUG, "r8168: Programming MAC Address...");

	/* Disable register protection */
	outb(CFG_9346_UNLOCK, io_base + CFG_9346);

	/* Set MAC address: only 4-byte write accesses allowed */
	outl(mac[4] | mac[5] << 8, io_base + 4);
	inl(io_base + 4);
	outl(mac[0] | mac[1] << 8 | mac[2] << 16 | mac[3] << 24,
		io_base);
	inl(io_base);
	/* Lock config regs */
	outb(CFG_9346_LOCK, io_base + CFG_9346);

	printk(BIOS_DEBUG, "done\n");
}

static void enable_aspm_l1_2(u16 io_base)
{
	printk(BIOS_INFO, "rtl: Enable ASPM L1.2\n");

	/* Disable register protection */
	outb(CFG_9346_UNLOCK, io_base + CFG_9346);

	/* Enable ASPM_L1.2 */
	outl(ASPM_L1_2_MASK, io_base + CMD_REG_ASPM);

	/* Lock config regs */
	outb(CFG_9346_LOCK, io_base + CFG_9346);
}

static void r8168_set_customized_led(struct device *dev, u16 io_base)
{
	struct drivers_net_config *config = dev->chip_info;

	if (!config)
		return;

	if (dev->device == PCI_DID_REALTEK_8125) {
		/* Set LED global Feature register */
		outb(config->led_feature, io_base + CMD_LED_FEATURE);
		printk(BIOS_DEBUG, "r8125: read back LED global feature setting as 0x%x\n",
		inb(io_base + CMD_LED_FEATURE));

		/*
		 * Refer to RTL8125 datasheet 5.Customizable LED Configuration
		 * Register Name	IO Address
		 * LEDSEL0		0x18
		 * LEDSEL2		0x84
		 * LEDFEATURE		0x94
		 *
		 * LEDSEL Bit[]		Description
		 * Bit0			Link10M
		 * Bit1			Link100M
		 * Bit3			Link1000M
		 * Bit5			Link2.5G
		 * Bit9			ACT
		 * Bit10		preboot enable
		 * Bit11		lp enable
		 * Bit12		active low/high
		 *
		 * LEDFEATURE		Description
		 * Bit0			LED Table V1/V2
		 * Bit1~3		Reserved
		 * Bit4~5		LED Blinking Duty Cycle	12.5%/ 25%/ 50%/ 75%
		 * Bit6~7		LED Blinking Freq. 240ms/160ms/80ms/Link-Speed-Dependent
		 */

		/* Set customized LED0 register */
		outw(config->customized_led0, io_base + CMD_LEDSEL0);
		printk(BIOS_DEBUG, "r8125: read back LED0 setting as 0x%x\n",
			inw(io_base + CMD_LEDSEL0));

		/* Set customized LED2 register */
		outw(config->customized_led2, io_base + CMD_LEDSEL2);
		printk(BIOS_DEBUG, "r8125: read back LED2 setting as 0x%x\n",
			inw(io_base + CMD_LEDSEL2));
	} else {
		/* Disable register protection */
		outb(CFG_9346_UNLOCK, io_base + CFG_9346);

		/* Read the customized LED setting from devicetree */
		printk(BIOS_DEBUG, "r8168: Customized LED 0x%x\n", config->customized_leds);

		/*
		 * Refer to RTL8111H datasheet 7.2 Customizable LED Configuration
		 * Starting from offset 0x18
		 * Bit[15:12]	LED Feature Control(FC)
		 * Bit[11:08]	LED Select for PINLED2
		 * Bit[07:04]	LED Select for PINLED1
		 * Bit[03:00]	LED Select for PINLED0
		 *
		 * Speed	Link10M		Link100M	Link1000M	ACT/Full
		 * LED0		Bit0		Bit1		Bit2		Bit3
		 * LED1		Bit4		Bit5		Bit6		Bit7
		 * LED2		Bit8		Bit9		Bit10		Bit11
		 * FC		Bit12		Bit13		Bit14		Bit15
		 */

		/* Enable customized LED mode */
		outb(io_base + 0x50, inb(io_base + 0x50) | 0x40);

		printk(BIOS_DEBUG, "r8168: current LED setting is 0x%x\n",
			inw(io_base + CMD_LED0_LED1));

		/* Set customized LED registers */
		outw(config->customized_leds, io_base+ CMD_LED0_LED1);
		printk(BIOS_DEBUG, "r8168: read back LED setting as 0x%x\n",
			inw(io_base + CMD_LED0_LED1));

		/* Lock config regs */
		outb(CFG_9346_LOCK, io_base + CFG_9346);
	}
}

static void r8168_init(struct device *dev)
{
	/* Get the resource of the NIC mmio */
	struct resource *nic_res = find_resource(dev, PCI_BASE_ADDRESS_0);
	u16 io_base = (u16)nic_res->base;

	/* Check if the base is invalid */
	if (!io_base) {
		printk(BIOS_ERR, "r8168: Error can't find IO resource\n");
		return;
	}
	/* Enable but do not set bus master */
	pci_write_config16(dev, PCI_COMMAND,
			   PCI_COMMAND_MEMORY | PCI_COMMAND_IO);

	/* Program MAC address based on CBFS "macaddress" containing
	 * a string AA:BB:CC:DD:EE:FF */
	if (CONFIG(RT8168_SET_MAC_ADDRESS))
		program_mac_address(dev, io_base);

	/* Program customized LED mode */
	if (CONFIG(RT8168_SET_LED_MODE))
		r8168_set_customized_led(dev, io_base);

	struct drivers_net_config *config = dev->chip_info;
	if (CONFIG(PCIEXP_ASPM) && config->enable_aspm_l1_2)
		enable_aspm_l1_2(io_base);
}

#if CONFIG(HAVE_ACPI_TABLES)
#define R8168_ACPI_HID "R8168"
static void r8168_net_fill_ssdt(const struct device *dev)
{
	struct drivers_net_config *config = dev->chip_info;
	const char *path = acpi_device_path(dev->bus->dev);
	u32 address;

	if (!path || !config)
		return;

	/* Device */
	acpigen_write_scope(path);
	acpigen_write_device(acpi_device_name(dev));
	acpigen_write_name_string("_HID", R8168_ACPI_HID);
	acpi_device_write_uid(dev);

	if (dev->chip_ops)
		acpigen_write_name_string("_DDN", dev->chip_ops->name);

	/* Power Resource */
	if (CONFIG(RT8168_GEN_ACPI_POWER_RESOURCE) && config->has_power_resource) {
		const struct acpi_power_res_params power_res_params = {
			.stop_gpio = &config->stop_gpio,
			.stop_delay_ms = config->stop_delay_ms,
			.stop_off_delay_ms = config->stop_off_delay_ms
		};
		acpi_device_add_power_res(&power_res_params);
	}

	/* Address */
	address = PCI_SLOT(dev->path.pci.devfn) & 0xffff;
	address <<= 16;
	address |= PCI_FUNC(dev->path.pci.devfn) & 0xffff;
	acpigen_write_name_dword("_ADR", address);

	/* Wake capabilities */
	if (config->wake)
		acpigen_write_PRW(config->wake, 3);

	acpigen_pop_len(); /* Device */
	acpigen_pop_len(); /* Scope */

	printk(BIOS_INFO, "%s.%s: %s %s\n", path, acpi_device_name(dev),
		dev->chip_ops ? dev->chip_ops->name : "", dev_path(dev));
}

static const char *r8168_net_acpi_name(const struct device *dev)
{
	return "RLTK";
}
#endif

static struct device_operations r8168_ops  = {
	.read_resources   = pci_dev_read_resources,
	.set_resources    = pci_dev_set_resources,
	.enable_resources = pci_dev_enable_resources,
	.init             = r8168_init,
#if CONFIG(HAVE_ACPI_TABLES)
	.acpi_name        = r8168_net_acpi_name,
	.acpi_fill_ssdt   = r8168_net_fill_ssdt,
#endif
};

static const unsigned short pci_device_ids[] = {
	PCI_DID_REALTEK_8168,
	PCI_DID_REALTEK_8125,
	PCI_DID_REALTEK_8111,
	0
};

static const struct pci_driver r8168_driver __pci_driver = {
	.ops    = &r8168_ops,
	.vendor = PCI_VID_REALTEK,
	.devices = pci_device_ids,
};

struct chip_operations drivers_net_ops = {
	CHIP_NAME("Realtek r8168")
};
